src/main/scala/com/exasol/spark/util/Filters.scala

```diff
@@ -10,99 +14,85 @@ import org.apache.spark.sql.types._
    */
-  def filterExpr(filter: Filter, dataTypes: Map[String, DataType]): Option[String] =
-    filter match {
-      case EqualTo(a, v)            => comparisonExpr(a, v, "=", dataTypes)
-      case Not(EqualTo(a, v))       => comparisonExpr(a, v, "!=", dataTypes)
-      case GreaterThan(a, v)        => comparisonExpr(a, v, ">", dataTypes)
-      case GreaterThanOrEqual(a, v) => comparisonExpr(a, v, ">=", dataTypes)
-      case LessThan(a, v)           => comparisonExpr(a, v, "<", dataTypes)
-      case LessThanOrEqual(a, v)    => comparisonExpr(a, v, "<=", dataTypes)
-      case IsNull(a)                => Option(s"($a IS NULL)")
-      case IsNotNull(a)             => Option(s"($a IS NOT NULL)")
-      case StringEndsWith(a, v)     => Option(s"($a LIKE '%$v')")
-      case StringContains(a, v)     => Option(s"($a LIKE '%$v%')")
-      case StringStartsWith(a, v)   => Option(s"($a LIKE '$v%')")
-      case In(a, vs)                => inExpr(a, vs, "IN", dataTypes)
-      case Not(In(a, vs))           => inExpr(a, vs, "NOT IN", dataTypes)
-      case Not(f)                   => notExpr(f, dataTypes)
-      case And(lf, rf)              => binaryExpr(lf, rf, "AND", dataTypes)
-      case Or(lf, rf)               => binaryExpr(lf, rf, "OR", dataTypes)
-      case _                        => None
-    }
-
-  def isQuotedValue(dataType: DataType, value: Any): String = dataType match {
-    case StringType    => s"'$value'"
-    case DateType      => s"date '$value'"
-    case TimestampType => s"timestamp '$value'"
-    case _             => s"$value"
-  }
-
-  def comparisonExpr(
-    name: String,
-    value: Any,
-    op: String,
-    types: Map[String, DataType]
-  ): Option[String] = types.get(name).flatMap {
-    case dtype =>
-      val newValue = isQuotedValue(dtype, value)
-      Option(s"$name $op $newValue")
-  }
-
-  def inExpr(
-    name: String,
-    values: Array[Any],
-    op: String,
-    types: Map[String, DataType]
-  ): Option[String] =
-    types.get(name).flatMap {
-      case dtype =>
-        val newValues = values.map(v => isQuotedValue(dtype, v))
-        Option(s"($name $op (${newValues.mkString(",")}))")
-    }
-
-  def notExpr(f: Filter, types: Map[String, DataType]): Option[String] =
-    filterExpr(f, types).flatMap(expr => Option(s"(NOT ($expr))"))
-
-  @SuppressWarnings(Array("org.wartremover.warts.Return", "org.wartremover.warts.OptionPartial"))
-  def binaryExpr(
-    lf: Filter,
-    rf: Filter,
-    op: String,
-    types: Map[String, DataType]
-  ): Option[String] = {
-    val leftOpt = filterExpr(lf, types)
-    if (!leftOpt.isDefined) {
-      return None // scalastyle:ignore return
-    }
+  // scalastyle:off null
+  // Suppression is accepted since we have terminal conditions in the
+  // recursion.
+  @SuppressWarnings(Array("org.wartremover.warts.Recursion"))
+  def filterToBooleanExpression(filter: Filter): Option[BooleanExpression] =
+    Option(filter match {
+      case EqualTo(attribute, value) =>
```

Comment:
wow, what a cool syntax. A very nice alternative to have visitors all the time.

Reply:
Yes, that is beauty of the pattern matching. 

It helps a lot if the language supports it as a language construct.